Watch mode doesn't correctly invalidate failed subdirectory lookup results

**TypeScript Version:**  2.8.0-dev.20180318

**Search Terms:** watch mode subdirectory

**Code**

You'll need 4 files
```ts
// tsconfig.json
// FOR THIS BUG, ORDER MATTERS IN THE FILE ARRAY
{
  "compilerOptions": {
    "target": "es5",
    "module": "commonjs"
  },
  "files": [
    "foo/a.ts",
    "foo/bar/b.ts"
  ]
}

// foo/a.ts
import * as q from 'qqq';

// foo/bar/b.ts
import * as q from 'qqq';

// node_modules/qqq/index_.d.ts
export { }
```

Run `tsc -w`:
```
20:37:52 - Starting compilation in watch mode...
foo/a.ts(1,20): error TS2307: Cannot find module 'qqq'.
foo/bar/b.ts(1,20): error TS2307: Cannot find module 'qqq'.
20:37:54 - Compilation complete. Watching for file changes.
```

Rename `index_.d.ts` to `index.d.ts`:
```
20:38:26 - File change detected. Starting incremental compilation...
foo/bar/b.ts(1,20): error TS2307: Cannot find module 'qqq'.
20:38:26 - Compilation complete. Watching for file changes.
```

**Expected behavior:** In the second compilation, `b.ts` should be able to correctly resolve `qqq`.

**Actual behavior:** Failure as shown above. Reversing the order of files in the `files` array produces the expected behavior.

**Playground Link:** N/A

**Related Issues:** None?

--- a/src/compiler/moduleNameResolver.ts
+++ b/src/compiler/moduleNameResolver.ts
@@ -1118,7 +1118,7 @@ namespace ts {
         const perModuleNameCache = cache && cache.getOrCreateCacheForModuleName(moduleName);
         return forEachAncestorDirectory(normalizeSlashes(directory), ancestorDirectory => {
             if (getBaseFileName(ancestorDirectory) !== "node_modules") {
-                const resolutionFromCache = tryFindNonRelativeModuleNameInCache(perModuleNameCache, moduleName, ancestorDirectory, state.traceEnabled, state.host);
+                const resolutionFromCache = tryFindNonRelativeModuleNameInCache(perModuleNameCache, moduleName, ancestorDirectory, state.traceEnabled, state.host, failedLookupLocations);
                 if (resolutionFromCache) {
                     return resolutionFromCache;
                 }
@@ -1196,12 +1196,13 @@ namespace ts {
             typesPackageName;
     }
 
-    function tryFindNonRelativeModuleNameInCache(cache: PerModuleNameCache | undefined, moduleName: string, containingDirectory: string, traceEnabled: boolean, host: ModuleResolutionHost): SearchResult<Resolved> {
+    function tryFindNonRelativeModuleNameInCache(cache: PerModuleNameCache | undefined, moduleName: string, containingDirectory: string, traceEnabled: boolean, host: ModuleResolutionHost, failedLookupLocations: Push<string>): SearchResult<Resolved> {
         const result = cache && cache.get(containingDirectory);
         if (result) {
             if (traceEnabled) {
                 trace(host, Diagnostics.Resolution_for_module_0_was_found_in_cache_from_location_1, moduleName, containingDirectory);
             }
+            failedLookupLocations.push(...result.failedLookupLocations);
             return { value: result.resolvedModule && { path: result.resolvedModule.resolvedFileName, extension: result.resolvedModule.extension, packageId: result.resolvedModule.packageId } };
         }
     }
@@ -1226,7 +1227,7 @@ namespace ts {
             if (!isExternalModuleNameRelative(moduleName)) {
                 // Climb up parent directories looking for a module.
                 const resolved = forEachAncestorDirectory(containingDirectory, directory => {
-                    const resolutionFromCache = tryFindNonRelativeModuleNameInCache(perModuleNameCache, moduleName, directory, traceEnabled, host);
+                    const resolutionFromCache = tryFindNonRelativeModuleNameInCache(perModuleNameCache, moduleName, directory, traceEnabled, host, failedLookupLocations);
                     if (resolutionFromCache) {
                         return resolutionFromCache;
                     }

--- a/src/harness/unittests/moduleResolution.ts
+++ b/src/harness/unittests/moduleResolution.ts
@@ -16,7 +16,7 @@ namespace ts {
     export function checkResolvedModuleWithFailedLookupLocations(actual: ResolvedModuleWithFailedLookupLocations, expectedResolvedModule: ResolvedModuleFull, expectedFailedLookupLocations: string[]): void {
         assert.isTrue(actual.resolvedModule !== undefined, "module should be resolved");
         checkResolvedModule(actual.resolvedModule, expectedResolvedModule);
-        assert.deepEqual(actual.failedLookupLocations, expectedFailedLookupLocations);
+        assert.deepEqual(actual.failedLookupLocations, expectedFailedLookupLocations, `Failed lookup locations should match - expected has ${expectedFailedLookupLocations.length}, actual has ${actual.failedLookupLocations.length}`);
     }
 
     export function createResolvedModule(resolvedFileName: string, isExternalLibraryImport = false): ResolvedModuleFull {

--- a/src/harness/unittests/tsserverProjectSystem.ts
+++ b/src/harness/unittests/tsserverProjectSystem.ts
@@ -6045,6 +6045,47 @@ namespace ts.projectSystem {
             });
         });
 
+        describe("Subfolder invalidations correctly include parent folder failed lookup locations QJWOP", () => {
+            it("Includes the parent folder FLLs", () => {
+                const projectLocation = "/proj";
+                const file1: FileOrFolder = {
+                    path: `${projectLocation}/foo/boo/app.ts`,
+                    content: `import * as debug from "debug"`
+                };
+                const file2: FileOrFolder = {
+                    path: `${projectLocation}/foo/boo/moo/app.ts`,
+                    content: `import * as debug from "debug"`
+                };
+                const tsconfig: FileOrFolder = {
+                    path: `${projectLocation}/tsconfig.json`,
+                    content: JSON.stringify({
+                        files: ["foo/boo/app.ts", "foo/boo/moo/app.ts"]
+                    })
+                };
+
+                const files = [file1, file2, tsconfig, libFile];
+                const host = createServerHost(files);
+                const service = createProjectService(host);
+                service.openClientFile(file1.path);
+
+                const project = service.configuredProjects.get(tsconfig.path);
+                checkProjectActualFiles(project, files.map(f => f.path));
+                assert.deepEqual(project.getLanguageService().getSemanticDiagnostics(file1.path).map(diag => diag.messageText), ["Cannot find module 'debug'."]);
+                assert.deepEqual(project.getLanguageService().getSemanticDiagnostics(file2.path).map(diag => diag.messageText), ["Cannot find module 'debug'."]);
+
+                const debugTypesFile: FileOrFolder = {
+                    path: `${projectLocation}/node_modules/debug/index.d.ts`,
+                    content: "export {}"
+                };
+                files.push(debugTypesFile);
+                host.reloadFS(files);
+                host.runQueuedTimeoutCallbacks();
+                checkProjectActualFiles(project, files.map(f => f.path));
+                assert.deepEqual(project.getLanguageService().getSemanticDiagnostics(file1.path).map(diag => diag.messageText), []);
+                assert.deepEqual(project.getLanguageService().getSemanticDiagnostics(file2.path).map(diag => diag.messageText), []);
+            });
+        });
+
         describe("Verify npm install in directory with tsconfig file works when", () => {
             function verifyNpmInstall(timeoutDuringPartialInstallation: boolean) {
                 const root = "/user/username/rootfolder/otherfolder";

--- a/src/harness/unittests/tsserverProjectSystem.ts
+++ b/src/harness/unittests/tsserverProjectSystem.ts
@@ -6045,8 +6045,8 @@ namespace ts.projectSystem {
             });
         });
 
-        describe("Subfolder invalidations correctly include parent folder failed lookup locations QJWOP", () => {
-            it("Includes the parent folder FLLs", () => {
+        describe("Subfolder invalidations correctly include parent folder failed lookup locations", () => {
+            function runFailedLookupTest(resolution: "Node" | "Classic") {
                 const projectLocation = "/proj";
                 const file1: FileOrFolder = {
                     path: `${projectLocation}/foo/boo/app.ts`,
@@ -6059,7 +6059,8 @@ namespace ts.projectSystem {
                 const tsconfig: FileOrFolder = {
                     path: `${projectLocation}/tsconfig.json`,
                     content: JSON.stringify({
-                        files: ["foo/boo/app.ts", "foo/boo/moo/app.ts"]
+                        files: ["foo/boo/app.ts", "foo/boo/moo/app.ts"],
+                        moduleResolution: resolution
                     })
                 };
 
@@ -6083,6 +6084,13 @@ namespace ts.projectSystem {
                 checkProjectActualFiles(project, files.map(f => f.path));
                 assert.deepEqual(project.getLanguageService().getSemanticDiagnostics(file1.path).map(diag => diag.messageText), []);
                 assert.deepEqual(project.getLanguageService().getSemanticDiagnostics(file2.path).map(diag => diag.messageText), []);
+            }
+
+            it("Includes the parent folder FLLs in node module resolution mode", () => {
+                runFailedLookupTest("Node");
+            });
+            it("Includes the parent folder FLLs in classic module resolution mode", () => {
+                runFailedLookupTest("Classic");
             });
         });
 
@@ -6251,6 +6259,44 @@ namespace ts.projectSystem {
             it("timeout occurs after installation", () => {
                 verifyNpmInstall(/*timeoutDuringPartialInstallation*/ false);
             });
+        });
+
+        it("when node_modules dont receive event for the @types file addition", () => {
+            const projectLocation = "/user/username/folder/myproject";
+            const app: FileOrFolder = {
+                path: `${projectLocation}/app.ts`,
+                content: `import * as debug from "debug"`
+            };
+            const tsconfig: FileOrFolder = {
+                path: `${projectLocation}/tsconfig.json`,
+                content: ""
+            };
+
+            const files = [app, tsconfig, libFile];
+            const host = createServerHost(files);
+            const service = createProjectService(host);
+            service.openClientFile(app.path);
+
+            const project = service.configuredProjects.get(tsconfig.path);
+            checkProjectActualFiles(project, files.map(f => f.path));
+            assert.deepEqual(project.getLanguageService().getSemanticDiagnostics(app.path).map(diag => diag.messageText), ["Cannot find module 'debug'."]);
+
+            const debugTypesFile: FileOrFolder = {
+                path: `${projectLocation}/node_modules/@types/debug/index.d.ts`,
+                content: "export {}"
+            };
+            files.push(debugTypesFile);
+            // Do not invoke recursive directory watcher for anything other than node_module/@types
+            const invoker = host.invokeWatchedDirectoriesRecursiveCallback;
+            host.invokeWatchedDirectoriesRecursiveCallback = (fullPath, relativePath) => {
+                if (fullPath.endsWith("@types")) {
+                    invoker.call(host, fullPath, relativePath);
+                }
+            };
+            host.reloadFS(files);
+            host.runQueuedTimeoutCallbacks();
+            checkProjectActualFiles(project, files.map(f => f.path));
+            assert.deepEqual(project.getLanguageService().getSemanticDiagnostics(app.path).map(diag => diag.messageText), []);
         });
     });
 

